### ISSUE TITLE
How to route audio to/from a macOS DAW?

### ISSUE THREAD

**Issue by bugeats** [VCVRack/Rack#532]
I've spent days now dicking around with Soundflower, JACK, Loopback, and whatever else I could find in a desperate attempt to sequence VCV Rack with virtual control voltage from my DAW (Bitwig on macOS). I have yet to find a solution that works.

**Soundflower** can be used as part of a native macOS Aggregate Device with my physical audio interface. However, Soundflower appears to always mirror output audio back to the input audio. In other words, patching to VCV Rack output 1 will mirror that signal on input 1. This causes all kinds of feedback issues. It becomes impossible to use Rack as a DAW "insert". Is this how Soundflower is meant to behave? I have no idea because there is no ownership of the project and no documentation. 

**JACK** is ... just holy hell what a mess. Initially there was some hope to use the "JACK-insert" VST as an insert in my DAW signal chain. This doesn't work however because the bundled 32-bit VST doesn't detect the 64-bit JACK server.  And then there's the small issue of no JACK support for Rack. Maybe there's some combo of Aggregate Devices and JACK virtual routing that can make this work, but I'll be damned if I could ever navigate my way there. 

**Loopback** only appears to support IO streams from native mac applications themselves, and doesn't appear to support any sort of virtual buses or multi channel audio. 

Ideally, I would just have a native CoreAudio device that represented a virtual audio bus. Soundflower is the closest to this, but again, it does this weird input/output mirroring thing. 

**TL;DR** Anybody out there figured out a way to use all 8 channels of the VCV Rack audio interface with their DAW?

**Comment by AlexanderMotylev** [comment]
Loopback supports virtual buses

<img width="771" alt="loopbacksettings" src="https://user-images.githubusercontent.com/34608193/34084122-63acafbe-e38c-11e7-861f-ef09bc0af0e2.png">

<img width="1440" alt="reaktor" src="https://user-images.githubusercontent.com/34608193/34084209-c3b44bf0-e38d-11e7-9da9-b1b28e1b86d4.png">

<img width="1440" alt="rack" src="https://user-images.githubusercontent.com/34608193/34084211-c942ee8c-e38d-11e7-84d5-4e25bd7eb4ea.png">

**Comment by bugeats** [comment]
Thank you @AlexanderMotylev. That's really helpful.

I ended up taking the time to sit down and try to understand JACK, and I eventually got to the point where I was able to complete my goal (despite general JACK bugginess) of routing everything to and from my hardware, DAW, and Rack interfaces. It's a pretty sweet setup now, actually.

If I wanted to write a guide on how to get JACK going with Rack, where would be the most appropriate place? The project doc files? The Github wiki?

**Comment by danballance** [comment]
@bugeats if you did have some time to write up your Jack configuration that would be massively helpful. I am running on Linux and have access to Jack but I have to admit I've not yet managed to put the time into getting a configuration together that works alongside my DAW (Bitwig).

**Comment by bugeats** [comment]
@danballance I dunno about Linux, but the macOS JACK distro has a native virtual audio device that can be assigned in VCV Rack. Once you assign it, that set of connections will show up in the JACK pool. From there you can name and assign ports in Bitwig.

It's all pretty obtuse and buggy, but I was able to get a setup working. I suggest taking the time to sit down and read the JACK docs.

**Comment by rossco255** [comment]
I've had success with Soundflower in Ableton. Using macOS's CoreAudio driver in Ableton allows you to have a separate device for input and output. Set the audio device in Rack as the 64ch Soundflower, then set the input device as the 64ch Soundflower and the output device as whatever you want to output from (soundcard, internal output, etc). Then it's just a case of setting up 8 audio channels in Ableton, each set to their respective inputs.

**Comment by n6smith** [comment]
FWIW... I gave up on both Jack and Soundflower due to inherent bugginess in both and have found Loopback to be the most stable.. Using that, i have successfully and reliablely sent audio to and from external i/o interfaces such as Motu's 16A and Ultralites Mk2, MK3 and AVS... AudioUnit plugins via hosts such as AU Lab and HU..  and so on without much of an issue.. Latency can be a little tricky but can be worked around depending on the config/patch used. I'll be publishing a video in the next week or so, demonstrating the techniques i have used to do so.. Ill update here with link when it is completed.

**Comment by genevera** [comment]
@bugeats You could try [Soundflower-2x4](https://github.com/kawauso/Soundflower-2x4). I use it to pipe sound from one Soundflower through `sox` and out to another.

**Comment by AndrewBelt** [comment]
Closing, since *VCV Rack for DAWs* will release shortly after Rack v2 is released, which will be the only supported way of using Rack in your DAW.
